src/Unity.Mathematics/quaternion.cs

```diff
@@ -629,7 +629,7 @@ public static quaternion log(quaternion q)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static quaternion mul(quaternion a, quaternion b)
         {
-            return quaternion(a.value.wwww * b.value + (a.value.xyzx * b.value.wwwx + a.value.yzxy * b.value.zxyy) * float4(1.0f, 1.0f, 1.0f, -1.0f) - a.value.zxyz * b.value.yzxz);
+            return chgsign((a.value.ywzx * b.value.xwyz - a.value.wxyz * b.value.zxzx - a.value.zzww * b.value.wzxy - a.value.xyxy * b.value.yyww).zwxy, float4(-1f, -1f, -1f, 1f));
```

Comment:
> Fixed quaternion mul to consider sign like core UnityEngine code.

@oldio is this the change you're referring to with the above comment?

I'm trying to understand what the difference is here, but when I worked out the expressions of both versions, I ended up with the same result which suggested to me that there is no difference.

Reply:
I would expect the result to be the same indeed, other than making sure that the multiplication logic happens on the same unit quaternion hemisphere (because of the chgsign).
@mecanimdev do you know if we have any tests in the core engine side that validates the chgsign usage in the quaternion mul operation that we could bring back to the package ?

Reply:
@unpacklo in terms of the swizzeling pattern here, we took what was already defined in the core engine SIMB math lib.

Reply:
just looked and unfortunately we don't have coverage for this one but it shouldn't be to hard to create a new one

Reply:
Here are some perf numbers from this change from my MacBook Pro which features an Intel i9-9880H CPU (my Windows desktop has similar numbers).

master:

```
quaternion_quaternion_burst (0.060s)
---
Burst Microsecond Median:218.90 Min:207.80 Max:278.50 Avg:226.31 Std:20.98 SampleCount: 10 Sum: 2263.10

quaternion_quaternion_mono (0.460s)
---
Mono Microsecond Median:34807.70 Min:34163.70 Max:36200.80 Avg:34949.11 Std:590.81 SampleCount: 10 Sum: 349491.10
```

This PR:

```
quaternion_quaternion_burst (0.060s)
---
Burst Microsecond Median:239.70 Min:237.00 Max:301.70 Avg:249.22 Std:21.39 SampleCount: 10 Sum: 2492.20

quaternion_quaternion_mono (0.804s)
---
Mono Microsecond Median:65296.00 Min:63279.90 Max:70754.60 Avg:66044.42 Std:2156.85 SampleCount: 10 Sum: 660444.20
```

```
burst baseline-new speedup: 0.908073188347645x
mono baseline-new speedup: 0.529175818335599x
```

The burst numbers look like it regressed but in my experience, this is well within the noise of the measurements and would require a lot more careful timings to see if there is actually a regression here. Mono is the main concern in this case, where it is now half the speed. The main reason for the performance regression is the usage of `chgsign`. I'll have to investigate if there's a way to speed up `chgsign` itself, but if we can't, is it OK for us to change this implementation to use a multiplication instead of `chgsign`?

`(a.value.ywzx * b.value.xwyz - a.value.wxyz * b.value.zxzx - a.value.zzww * b.value.wzxy - a.value.xyxy * b.value.yyww).zwxy * float4(-1f, -1f, -1f, 1f);`

This yields performance numbers that look like this:

```
quaternion_quaternion_burst (0.080s)
---
Burst Microsecond Median:285.60 Min:282.50 Max:324.50 Avg:288.75 Std:12.00 SampleCount: 10 Sum: 2887.50

quaternion_quaternion_mono (0.500s)
---
Mono Microsecond Median:38627.40 Min:37820.50 Max:42265.50 Avg:39073.62 Std:1351.92 SampleCount: 10 Sum: 390736.20
```

The speedup from master to this version in mono is: 0.894442593238098x

Reply:
Doing the mul instead of chgsign should be fine imo. @mecanimdev any concerns on your side ?

Reply:
Otherwise just saw your comment @unpacklo : https://unity.slack.com/archives/CQVNHR5R8/p1637631914045100
That probably has a significant impact on `chgsign`...

Reply:
no, in both code path we are doing the conjugate of the quaternion.

Let's take the one that generate less instructions and cpu cycle. I would be curious to compare the assembly version of Unity.Mathematics and c++ simd math library.

Reply:
> Otherwise just saw your comment @unpacklo : https://unity.slack.com/archives/CQVNHR5R8/p1637631914045100 That probably has a significant impact on `chgsign`. Good find!

Yep! It's huge right now, I don't have final numbers yet but for just converting from float4 to uint4, I saw a number last night that was 20x speedup!!! I've already got a separate branch with some of those changes that I intend to land separately but I'm definitely going to check this PR with those changes and report back on the new perf.

Reply:
Finally managed to get new perf numbers now that I landed #206. Here are the numbers in my performance test from my Windows desktop which has an Intel Core i7-7820X (timings are in microseconds for 500k quaternion-quaternion multiplies):

| Method                      | Mono baseline | Mono new  | Mono speedup | Burst baseline | Burst new | Burst speedup |
| --------------------------- | ------------- | --------- | ------------ | -------------- | --------- | ------------- |
| mul(quaternion, quaternion) | 965251.1      | 1098053.8 | 0.8790562903 | 15509.6        | 16560.7   | 0.9365304607  |

I'm currently in favor of keeping the baseline (master branch version) of `mul(quaternion, quaternion)` implementation due to the higher performance.

@mecanimdev Here are the code listings for various versions of `mul(quaternion, quaternion)`.

Baseline (master branch version):

```asm
.LBB0_1:                                # %BL.0004.i
                                        # =>This Inner Loop Header: Depth=1
.Ltmp0:
        .loc        1 242 49 prologue_end   # TestMul.gen.cs:242:49
        mov        rdx, qword ptr [rcx + 8]
        mov        r8, qword ptr [rcx + 16]
        movups        xmm1, xmmword ptr [rdx + rax + 8000000]
        movups        xmm2, xmmword ptr [r8 + rax + 8000000]
.Ltmp1:
        .file        2 "C:/Users/dale/code/Unity.Mathematics/src/Unity.Mathematics\\quaternion.cs"
        .loc        2 632 13                # quaternion.cs:632:13
        movaps        xmm3, xmm1
        shufps        xmm3, xmm1, 255         # xmm3 = xmm3[3,3],xmm1[3,3]
        mulps        xmm3, xmm2
        movaps        xmm4, xmm1
        shufps        xmm4, xmm1, 36          # xmm4 = xmm4[0,1],xmm1[2,0]
        movaps        xmm5, xmm2
        shufps        xmm5, xmm2, 63          # xmm5 = xmm5[3,3],xmm2[3,0]
        mulps        xmm5, xmm4
        movaps        xmm4, xmm1
        shufps        xmm4, xmm1, 73          # xmm4 = xmm4[1,2],xmm1[0,1]
        movaps        xmm6, xmm2
        shufps        xmm6, xmm2, 82          # xmm6 = xmm6[2,0],xmm2[1,1]
        mulps        xmm6, xmm4
        addps        xmm6, xmm5
        mulps        xmm6, xmm0
        addps        xmm6, xmm3
        shufps        xmm1, xmm1, 146         # xmm1 = xmm1[2,0,1,2]
        shufps        xmm2, xmm2, 137         # xmm2 = xmm2[1,2,0,2]
        mulps        xmm2, xmm1
        subps        xmm6, xmm2
.Ltmp2:
        .loc        1 242 49                # TestMul.gen.cs:242:49
        movups        xmmword ptr [r8 + rax + 8000000], xmm6
        .loc        1 246 13                # TestMul.gen.cs:246:13
        add        rax, 16
        jne        .LBB0_1
```

The new mul from this PR:

```asm
.LBB0_1:                                # %BL.0004.i
                                        # =>This Inner Loop Header: Depth=1
.Ltmp0:
        .loc        1 242 49 prologue_end   # TestMul.gen.cs:242:49
        mov        rdx, qword ptr [rcx + 8]
        mov        r8, qword ptr [rcx + 16]
        movups        xmm1, xmmword ptr [rdx + rax + 8000000]
        movups        xmm2, xmmword ptr [r8 + rax + 8000000]
.Ltmp1:
        .file        2 "C:/Users/dale/code/Unity.Mathematics/src/Unity.Mathematics\\quaternion.cs"
        .loc        2 632 13                # quaternion.cs:632:13
        movaps        xmm3, xmm1
        shufps        xmm3, xmm1, 45          # xmm3 = xmm3[1,3],xmm1[2,0]
        movaps        xmm4, xmm2
        shufps        xmm4, xmm2, 156         # xmm4 = xmm4[0,3],xmm2[1,2]
        mulps        xmm4, xmm3
        movaps        xmm3, xmm1
        shufps        xmm3, xmm1, 147         # xmm3 = xmm3[3,0],xmm1[1,2]
        movaps        xmm5, xmm2
        shufps        xmm5, xmm2, 34          # xmm5 = xmm5[2,0],xmm2[2,0]
        mulps        xmm5, xmm3
        subps        xmm4, xmm5
        movddup        xmm3, xmm1              # xmm3 = xmm1[0,0]
        unpckhps        xmm1, xmm1      # xmm1 = xmm1[2,2,3,3]
        movshdup        xmm5, xmm2      # xmm5 = xmm2[1,1,3,3]
        shufps        xmm2, xmm2, 75          # xmm2 = xmm2[3,2,0,1]
        mulps        xmm2, xmm1
        subps        xmm4, xmm2
        mulps        xmm5, xmm3
        subps        xmm4, xmm5
        shufpd        xmm4, xmm4, 1           # xmm4 = xmm4[1,0]
.Ltmp2:
        .file        3 "C:/Users/dale/code/Unity.Mathematics/src/Unity.Mathematics\\math.cs"
        .loc        3 6640 13               # math.cs:6640:13
        xorpd        xmm4, xmm0
.Ltmp3:
        .loc        1 242 49                # TestMul.gen.cs:242:49
        movupd        xmmword ptr [r8 + rax + 8000000], xmm4
        .loc        1 246 13                # TestMul.gen.cs:246:13
        add        rax, 16
        jne        .LBB0_1
```

And the disassembled code for `quatMul` in our C++ simd library:

```asm
--- C:\Users\dale\code\unity\Runtime\Math\Simd\vec-quat.cpp --------------------
        return quatMul(*q1, *q2);
00007FF6A714CFE0  movups      xmm5,xmmword ptr [rdx]  
00007FF6A714CFE3  mov         rax,rcx  
00007FF6A714CFE6  movups      xmm3,xmmword ptr [r8]  
00007FF6A714CFEA  movaps      xmm4,xmm5  
00007FF6A714CFED  movlhps     xmm4,xmm5  
00007FF6A714CFF0  movaps      xmm1,xmm5  
00007FF6A714CFF3  pshufd      xmm0,xmm3,0F5h  
00007FF6A714CFF8  movaps      xmm2,xmm3  
00007FF6A714CFFB  mulps       xmm4,xmm0  
00007FF6A714CFFE  pshufd      xmm0,xmm5,0AFh  
00007FF6A714D003  mulps       xmm2,xmm0  
00007FF6A714D006  pshufd      xmm0,xmm3,88h  
00007FF6A714D00B  mulps       xmm1,xmm0  
00007FF6A714D00E  pshufd      xmm0,xmm3,0D2h  
00007FF6A714D013  mulps       xmm5,xmm0  
00007FF6A714D016  pshufd      xmm0,xmm1,1Eh  
00007FF6A714D01B  subps       xmm5,xmm0  
00007FF6A714D01E  pshufd      xmm0,xmm2,8Dh  
00007FF6A714D023  subps       xmm5,xmm0  
00007FF6A714D026  pshufd      xmm0,xmm4,63h  
00007FF6A714D02B  subps       xmm5,xmm0  
00007FF6A714D02E  movdqa      xmm0,xmmword ptr [__xmm@00000000800000008000000080000000 (07FF6AA18BA40h)]  
00007FF6A714D036  pshufd      xmm1,xmm5,0D2h  
00007FF6A714D03B  pxor        xmm0,xmm1  
00007FF6A714D03F  movdqu      xmmword ptr [rcx],xmm0  
    }
00007FF6A714D043  ret  
```

Reply:
yeah seem legit to keep the current one if they both yield the same result

Reply:
@sschoener mentioned elsewhere that the perf difference of around 7% for burst could be due to noise, so I ran some t-tests:

![image](https://user-images.githubusercontent.com/3121968/146474733-9426e832-ee47-4030-ac2b-0d90a0fb6e82.png)

The test runs are total times the performance test framework reports for the 500k quat muls performed 10 times (1 warmup is done before measurement). Mac is a bit interesting because I'm on a MacBook Pro and I usually see wildly varying times with the default settings (whatever that is for CPU throttling and such), so I had to do both with the default settings and another with Intel Power Gadget running in the background set to maximum CPU frequency.

![image](https://user-images.githubusercontent.com/3121968/146474040-f3cf8b94-0e2c-4b9c-af12-522aec8e33d6.png)

`Mac baseline vs new burst` is the only one that fails the statistical significance test and we cannot make a conclusion about whether or not the baseline perf and new perf are different. However, all other comparisons show that there is a statistically significant difference between baseline and new. Despite the fact that the `Mac baseline vs new burst` t-test failed to show a statistically significant difference, I feel confident in concluding that the baseline version is still better because the `Mac max freq baseline vs new burst` test is running the exact same code, the main difference being a background application running to force the CPU to stay at a higher frequency.

Finally, I was curious to know why the baseline version was faster and now I think I have a clue as to why. I found [uiCA](https://uica.uops.info/) which simulates an assembly instruction sequence and analyzes bottlenecks and here are links to the various versions of the code:

[Baseline Burst](https://uica.uops.info/?code=.LBB0_1%3A%0D%0A.Ltmp0%3A%0D%0A%20%20%20%20%20%20%20%20mov%20%20%20%20%20%20%20%20rdx%2C%20qword%20ptr%20%5Brcx%20%2B%208%5D%0D%0A%20%20%20%20%20%20%20%20mov%20%20%20%20%20%20%20%20r8%2C%20qword%20ptr%20%5Brcx%20%2B%2016%5D%0D%0A%20%20%20%20%20%20%20%20movups%20%20%20%20%20%20%20%20xmm1%2C%20xmmword%20ptr%20%5Brdx%20%2B%20rax%20%2B%208000000%5D%0D%0A%20%20%20%20%20%20%20%20movups%20%20%20%20%20%20%20%20xmm2%2C%20xmmword%20ptr%20%5Br8%20%2B%20rax%20%2B%208000000%5D%0D%0A.Ltmp1%3A%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm3%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm3%2C%20xmm1%2C%20255%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm3%2C%20xmm2%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm4%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm4%2C%20xmm1%2C%2036%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm5%2C%20xmm2%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm5%2C%20xmm2%2C%2063%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm5%2C%20xmm4%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm4%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm4%2C%20xmm1%2C%2073%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm6%2C%20xmm2%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm6%2C%20xmm2%2C%2082%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm6%2C%20xmm4%0D%0A%20%20%20%20%20%20%20%20addps%20%20%20%20%20%20%20%20xmm6%2C%20xmm5%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm6%2C%20xmm0%0D%0A%20%20%20%20%20%20%20%20addps%20%20%20%20%20%20%20%20xmm6%2C%20xmm3%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm1%2C%20xmm1%2C%20146%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm2%2C%20xmm2%2C%20137%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm2%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20subps%20%20%20%20%20%20%20%20xmm6%2C%20xmm2%0D%0A.Ltmp2%3A%0D%0A%0D%0A%20%20%20%20%20%20%20%20movups%20%20%20%20%20%20%20%20xmmword%20ptr%20%5Br8%20%2B%20rax%20%2B%208000000%5D%2C%20xmm6%0D%0A%20%20%20%20%20%20%20%20add%20%20%20%20%20%20%20%20rax%2C%2016%0D%0A%20%20%20%20%20%20%20%20jne%20%20%20%20%20%20%20%20.LBB0_1%0D%0A&syntax=asIntel&uArchs=SKL&uArchs=SKX&uArchs=TGL&uArchs=RKL&tools=uiCA&tools=llvm&alignment=0)
![image](https://user-images.githubusercontent.com/3121968/146476605-95f3bb60-05e9-4d5a-9482-946d93e4dc30.png)

[New Burst](https://uica.uops.info/?code=.LBB0_1%3A%0D%0A.Ltmp0%3A%0D%0A%20%20%20%20%20%20%20%20mov%20%20%20%20%20%20%20%20rdx%2C%20qword%20ptr%20%5Brcx%20%2B%208%5D%0D%0A%20%20%20%20%20%20%20%20mov%20%20%20%20%20%20%20%20r8%2C%20qword%20ptr%20%5Brcx%20%2B%2016%5D%0D%0A%20%20%20%20%20%20%20%20movups%20%20%20%20%20%20%20%20xmm1%2C%20xmmword%20ptr%20%5Brdx%20%2B%20rax%20%2B%208000000%5D%0D%0A%20%20%20%20%20%20%20%20movups%20%20%20%20%20%20%20%20xmm2%2C%20xmmword%20ptr%20%5Br8%20%2B%20rax%20%2B%208000000%5D%0D%0A.Ltmp1%3A%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm3%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm3%2C%20xmm1%2C%2045%20%20%20%20%20%20%20%20%20%20%23%20xmm3%20%3D%20xmm3%5B1%2C3%5D%2Cxmm1%5B2%2C0%5D%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm4%2C%20xmm2%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm4%2C%20xmm2%2C%20156%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm4%2C%20xmm3%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm3%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm3%2C%20xmm1%2C%20147%0D%0A%20%20%20%20%20%20%20%20movaps%20%20%20%20%20%20%20%20xmm5%2C%20xmm2%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm5%2C%20xmm2%2C%2034%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm5%2C%20xmm3%0D%0A%20%20%20%20%20%20%20%20subps%20%20%20%20%20%20%20%20xmm4%2C%20xmm5%0D%0A%20%20%20%20%20%20%20%20movddup%20%20%20%20%20%20%20%20xmm3%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20unpckhps%20%20%20%20%20%20%20%20xmm1%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20movshdup%20%20%20%20%20%20%20%20xmm5%2C%20xmm2%0D%0A%20%20%20%20%20%20%20%20shufps%20%20%20%20%20%20%20%20xmm2%2C%20xmm2%2C%2075%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm2%2C%20xmm1%0D%0A%20%20%20%20%20%20%20%20subps%20%20%20%20%20%20%20%20xmm4%2C%20xmm2%0D%0A%20%20%20%20%20%20%20%20mulps%20%20%20%20%20%20%20%20xmm5%2C%20xmm3%0D%0A%20%20%20%20%20%20%20%20subps%20%20%20%20%20%20%20%20xmm4%2C%20xmm5%0D%0A%20%20%20%20%20%20%20%20shufpd%20%20%20%20%20%20%20%20xmm4%2C%20xmm4%2C%201%0D%0A.Ltmp2%3A%0D%0A%20%20%20%20%20%20%20%20xorpd%20%20%20%20%20%20%20%20xmm4%2C%20xmm0%0D%0A.Ltmp3%3A%0D%0A%20%20%20%20%20%20%20%20movupd%20%20%20%20%20%20%20%20xmmword%20ptr%20%5Br8%20%2B%20rax%20%2B%208000000%5D%2C%20xmm4%0D%0A%20%20%20%20%20%20%20%20add%20%20%20%20%20%20%20%20rax%2C%2016%0D%0A%20%20%20%20%20%20%20%20jne%20%20%20%20%20%20%20%20.LBB0_1&syntax=asIntel&uArchs=SKL&uArchs=SKX&uArchs=TGL&uArchs=RKL&tools=uiCA&tools=llvm&alignment=0)
![image](https://user-images.githubusercontent.com/3121968/146476637-744859fd-b437-4509-bc52-08bc857b873f.png)

It appears that the reason why the baseline version is faster is due to a shorter instruction sequence and less shuffling done on port 5, whereas the new version is slightly longer and has even more work done on port 5.

As a bonus, I also mocked up a similar loop with the C++ simd code we have and dumped the assembly into uiCA (some massaging of the assembly was required for uiCA to accept it):

```c++
    void doStuff(int n, const float4* q1, const float4* q2, float4* results)
    {
        for (int i = 0; i < n; ++i)
        {
            results[i] = quatMul(q1[i], q2[i]);
        }
    }
```

```asm
Unity.exe!math::doStuff(int, const math::_float4 *, const math::_float4 *, const math::_float4 *):
00007FF7D06E1880  sub         rsp,18h  
00007FF7D06E1884  test        ecx,ecx  
00007FF7D06E1886  jle         math::doStuff+83h (07FF7D06E1903h)  
00007FF7D06E1888  sub         r8,rdx  
00007FF7D06E188B  movaps      xmmword ptr [rsp],xmm6  
00007FF7D06E188F  movdqa      xmm6,xmmword ptr [__xmm@00000000800000008000000080000000 (07FF7DBA9BA40h)]  
00007FF7D06E1897  sub         r9,rdx  
00007FF7D06E189A  mov         eax,ecx  
00007FF7D06E189C  nop         dword ptr [rax]  
00007FF7D06E18A0  movups      xmm1,xmmword ptr [rdx]  
00007FF7D06E18A3  movups      xmm2,xmmword ptr [r8+rdx]  
00007FF7D06E18A8  movaps      xmm5,xmm1  
00007FF7D06E18AB  pshufd      xmm0,xmm2,0F5h  
00007FF7D06E18B0  movlhps     xmm5,xmm1  
00007FF7D06E18B3  mulps       xmm5,xmm0  
00007FF7D06E18B6  pshufd      xmm0,xmm2,88h  
00007FF7D06E18BB  mulps       xmm0,xmm1  
00007FF7D06E18BE  pshufd      xmm4,xmm2,0D2h  
00007FF7D06E18C3  mulps       xmm4,xmm1  
00007FF7D06E18C6  pshufd      xmm0,xmm0,1Eh  
00007FF7D06E18CB  pshufd      xmm3,xmm1,0AFh  
00007FF7D06E18D0  mulps       xmm3,xmm2  
00007FF7D06E18D3  subps       xmm4,xmm0  
00007FF7D06E18D6  pshufd      xmm0,xmm3,8Dh  
00007FF7D06E18DB  subps       xmm4,xmm0  
00007FF7D06E18DE  pshufd      xmm0,xmm5,63h  
00007FF7D06E18E3  subps       xmm4,xmm0  
00007FF7D06E18E6  pshufd      xmm0,xmm4,0D2h  
00007FF7D06E18EB  pxor        xmm0,xmm6  
00007FF7D06E18EF  movdqu      xmmword ptr [r9+rdx],xmm0  
00007FF7D06E18F5  add         rdx,10h  
00007FF7D06E18F9  sub         rax,1  
00007FF7D06E18FD  jne         math::doStuff+20h (07FF7D06E18A0h)  
00007FF7D06E18FF  movaps      xmm6,xmmword ptr [rsp]  
00007FF7D06E1903  add         rsp,18h  
00007FF7D06E1907  ret  
```
[C++ simd](https://bit.ly/3scUGyb)
![image](https://user-images.githubusercontent.com/3121968/146479356-8b9a77dc-d21e-479e-891c-fef69caa2a93.png)

The bottlenecks appear to be similar as before, but the throughput of the baseline Mathematics code still appears to be better and leads me to suspect that it is the fastest version of the three we are considering here.

Reply:
And finally, here's the simplified expressions of both implementations, which shows they are equivalent:

Baseline:
```
Assume a = (ax, ay, az, aw) and b = (bx, by, bz, bw).

Original expression:
quaternion(a.value.wwww * b.value + (a.value.xyzx * b.value.wwwx + a.value.yzxy * b.value.zxyy) * float4(1.0f, 1.0f, 1.0f, -1.0f) - a.value.zxyz * b.value.yzxz)

Simplified:
((aw, aw, aw, aw) * (bx, by, bz, bw)) + (((ax, ay, az, ax) * (bw, bw, bw, bx) + (ay, az, ax, ay) * (bz, bx, by, by)) * (1, 1, 1, -1)) - ((az, ax, ay, az) * (by, bz, bx, bz))
(aw * bx, aw * by, aw * bz, aw * bw) + (((ax * bw, ay * bw, az * bw, ax * bx) + (ay * bz, az * bx, ax * by, ay * by)) * (1, 1, 1, -1)) - (az * by, ax * bz, ay * bx, az * bz)
(aw * bx, aw * by, aw * bz, aw * bw) + ((ax * bw + ay * bz, ay * bw + az * bx, az * bw + ax * by, ax * bx + ay * by) * (1, 1, 1, -1)) - (az * by, ax * bz, ay * bx, az * bz)
(aw * bx, aw * by, aw * bz, aw * bw) + (ax * bw + ay * bz, ay * bw + az * bx, az * bw + ax * by, -ax * bx - ay * by) - (az * by, ax * bz, ay * bx, az * bz)
(aw * bx + ax * bw + ay * bz, aw * by + ay * bw + az * bx, aw * bz + az * bw + ax * by, aw * bw - ax * bx - ay * by) - (az * by, ax * bz, ay * bx, az * bz)
(aw * bx + ax * bw + ay * bz - az * by, aw * by + ay * bw + az * bx - ax * bz, aw * bz + az * bw + ax * by - ay * bx, aw * bw - ax * bx - ay * by - az * bz)

(aw * bx + ax * bw + ay * bz - az * by,
 aw * by + ay * bw + az * bx - ax * bz,
 aw * bz + az * bw + ax * by - ay * bx,
 aw * bw - ax * bx - ay * by - az * bz)
```

New:
```
Assume a = (ax, ay, az, aw) and b = (bx, by, bz, bw).

Original expression:
chgsign((a.value.ywzx * b.value.xwyz - a.value.wxyz * b.value.zxzx - a.value.zzww * b.value.wzxy - a.value.xyxy * b.value.yyww).zwxy, float4(-1f, -1f, -1f, 1f))

Simplified:
chgsign((((ay, aw, az, ax) * (bx, bw, by, bz)) - ((aw, ax, ay, az) * (bz, bx, bz, bx)) - ((az, az, aw, aw) * (bw, bz, bx, by)) - ((ax, ay, ax, ay) * (by, by, bw, bw))).zwxy, (-1, -1, -1, 1))
chgsign(((ay * bx, aw * bw, az * by, ax * bz) - (aw * bz, ax * bx, ay * bz, az * bx) - (az * bw, az * bz, aw * bx, aw * by) - (ax * by, ay * by, ax * bw, ay * bw)).zwxy, (-1, -1, -1, 1))
chgsign(((ay * bx - aw * bz, aw * bw - ax * bx, az * by - ay * bz, ax * bz - az * bx) - (az * bw, az * bz, aw * bx, aw * by) - (ax * by, ay * by, ax * bw, ay * bw)).zwxy, (-1, -1, -1, 1))
chgsign(((ay * bx - aw * bz - az * bw, aw * bw - ax * bx - az * bz, az * by - ay * bz - aw * bx, ax * bz - az * bx - aw * by) - (ax * by, ay * by, ax * bw, ay * bw)).zwxy, (-1, -1, -1, 1))
chgsign((ay * bx - aw * bz - az * bw - ax * by, aw * bw - ax * bx - az * bz - ay * by, az * by - ay * bz - aw * bx - ax * bw, ax * bz - az * bx - aw * by - ay * bw).zwxy, (-1, -1, -1, 1))
chgsign((az * by - ay * bz - aw * bx - ax * bw, ax * bz - az * bx - aw * by - ay * bw, ay * bx - aw * bz - az * bw - ax * by, aw * bw - ax * bx - az * bz - ay * by), (-1, -1, -1, 1))
(-az * by + ay * bz + aw * bx + ax * bw, -ax * bz + az * bx + aw * by + ay * bw, -ay * bx + aw * bz + az * bw + ax * by, aw * bw - ax * bx - az * bz - ay * by)
(aw * bx + ax * bw + ay * bz - az * by, aw * by + ay * bw + az * bx - ax * bz, aw * bz + az * bw + ax * by - ay * bx, aw * bw - ax * bx - ay * by - az * bz)

(aw * bx + ax * bw + ay * bz - az * by,
 aw * by + ay * bw + az * bx - ax * bz,
 aw * bz + az * bw + ax * by - ay * bx,
 aw * bw - ax * bx - ay * by - az * bz)
```